### PR TITLE
[FLINK-19494][DataStream API] Adjust StreamExecutionEnvironment.generateSequence() to new API Sources

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently the utility method `StreamExecutionEnvironment.generateSequence()` instantiates the `StatefulSequenceSource`. `StreamExecutionEnvironment` should add a new `fromSequence()` method that uses the `NumberSequenceSource` against the new `Source` interface based on FLIP-27.*

## Brief change log

  - *`StreamExecutionEnvironment` adds `fromSequence()` method to creates a new data stream that contains a sequence of numbers.*

## Verifying this change

  - *`StreamExecutionEnvironmentTest` add `testFromSequence()` test method to verify whether `fromSequence()` instantiates a new DataStream that contains a sequence of numbers.*
  - *Modify `SourceNAryInputChainingITCase` to use `fromSequence()` method instead of `fromSource(NumberSequenceSource)` to verify whether `fromSource(NumberSequenceSource)` could be replaced with `fromSequence()`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@StephanEwen @aljoscha Could you please help to review this adjust?

**Comment by SteNicholas** [comment]
@flinkbot run azure

**Comment by SteNicholas** [comment]
@StephanEwen @aljoscha Could you please help to review this adjust?

**Comment by StephanEwen** [comment]
Looks good to me, in general, with a few small follow-ups:

  1. I think the JavaDocs for the method could be improved slightly. Maybe you can just reuse the JavaDocs from the source implementation itself? https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java

  2. We should also deprecate the old method (`generateSequence()`) so that users are not confused which method to use.

  3. The `fromSource()` method takes the name directly as an explicit parameter. I am wondering if we should adopt that style also in the new `fromSequence()` method. @aljoscha what do you think?

**Comment by SteNicholas** [comment]
@StephanEwen 

- About the JavaDocs, I have integrated JavaDocs of method `generateSequence` and JavaDocs of class `NumberSequenceSource`. 
- if `generateSequence()` should be deprecated, I think that method `addSource()` should also be deprecated. I didn't deprecate the `generateSequence()` method because the corresponding `addSource()` method isn't deprecated.

I would like to improve the JavaDocs of `fromSequence` method more slightly.

**Comment by StephanEwen** [comment]
`generateSequence()` behaves a bit different, especially when it comes to order, so mixing the JavaDocs may be tricky. I would personally rather stay closer to the docs of the new `NumberSequenceSource`.

I think there is no reason to tie deprecation of `addSource()` and `generateSequence()` together.
We cannot deprecate `addSource()` because many sources still need it, but we also should not have two source methods for the same thing (`generateSequence()` and `fromSequence()`).

**Comment by SteNicholas** [comment]
@StephanEwen Thanks for your detailed explanation. I have already followed the comments you mentioned. The remaining method naming question for `fromSequence()` need @aljoscha 's suggestion.

**Comment by SteNicholas** [comment]
@StephanEwen ,thanks for your detailed explanation. I have followed the comments you mentioned to update the JavaDoc.
@aljoscha , what do you think about the naming of `fromSequence()` method?

**Comment by StephanEwen** [comment]
A few more comments on this:

  - `@Experimental` annotations should not just be added lightly. If we deprecate one method (`generateSequence()`) then the replacement for that should not be experimental, because that leaves users wondering what they should use then.

  - The Scala API should her not be tested with a test that executed the program. That is not very targeted testing. The Scala API only wraps the source, which can be tested in a much more lightweight way. Anything that an execution-based test verifies is relevant to the main source, not the Scala wrapper.

I am a bit unsure how we want to do watermarks in the `fromSequence()` case. Should we have no watermarks by default? I think we should at least have an option that supports adding timestamps/watermarks. 
@aljoscha  what is your take here?

**Comment by SteNicholas** [comment]
> A few more comments on this:
> 
> * `@Experimental` annotations should not just be added lightly. If we deprecate one method (`generateSequence()`) then the replacement for that should not be experimental, because that leaves users wondering what they should use then.
> * The Scala API should her not be tested with a test that executed the program. That is not very targeted testing. The Scala API only wraps the source, which can be tested in a much more lightweight way. Anything that an execution-based test verifies is relevant to the main source, not the Scala wrapper.
> 
> I am a bit unsure how we want to do watermarks in the `fromSequence()` case. Should we have no watermarks by default? I think we should at least have an option that supports adding timestamps/watermarks.
> @aljoscha what is your take here?

@StephanEwen , thanks for your comments above mentioned. Previously I thought that `fromSequence()` should be consistent with `fromSource()` so that `fromSequence()` would be added with `@Experimental` annotations. I have followed up with the comments by removing the annotation and updated the Scala API test cases. Please review again if you are available. Sorry for trouble you again.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 1]
```diff
@@ -832,14 +833,42 @@ private StateBackend loadStateBackend(ReadableConfig configuration, ClassLoader
 	 * @param to
 	 * 		The number to stop at (inclusive)
 	 * @return A data stream, containing all number in the [from, to] interval
+	 * @deprecated Use {@link #fromSequence(long, long)} instead to create a new data stream
+	 * that contains {@link org.apache.flink.api.connector.source.lib.NumberSequenceSource}.
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:837]: We shouldn't mention the internal implementation here. 

```suggestion
	 * @deprecated Use {@link #fromSequence(long, long)} instead to create a new sequence data stream.
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 2]
```diff
@@ -832,14 +833,42 @@ private StateBackend loadStateBackend(ReadableConfig configuration, ClassLoader
...
+	@Deprecated
 	public DataStreamSource<Long> generateSequence(long from, long to) {
 		if (from > to) {
 			throw new IllegalArgumentException("Start of sequence must not be greater than the end");
 		}
 		return addSource(new StatefulSequenceSource(from, to), "Sequence Source");
 	}
 
+	/**
+	 * Creates a new data stream that contains {@link org.apache.flink.api.connector.source.lib.NumberSequenceSource}
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:848]: I don't think we need to mention that implementation detail here.

@StephanEwen Sorry if you requested that this be mentioned here, then we can leave it in. 😅

**Comment by aljoscha** [comment]
Thanks for the heads up!

Regarding operator naming, I think it's good to have the name in the method, we just need to decide on a position and be consistent with it. I personally would prefer `fromSource(name, other, stuff)` because it keeps the name closer to the beginning of the parameters, which will become easier to read when you extend naming to things like `map(name, function)`, where the function can be a lambda. But I'm fine with either position, i.e. name at the beginning or the end. I just want consistency. 😅

Regarding watermarks: I would not add an overloaded method that takes a watermark strategy. If users need that they can use `fromSource()` and I think if they need custom watermarking they probably can use that "lower level" method. What we could do is use the `Long` itself as the timestamp and use it for watermarks. I think that would be the only sensible basis for a timestamp/watermark that is deterministic. What do you think?

**Comment by StephanEwen** [comment]
### About Timestamps / Watermarks

@aljoscha So if anyone would need timestamps/watermarks, they would need to manually create the `NumberSequenceSource`?
I am wondering if in that case, we should just remove the "convenience method" as a whole?

### About Name

@aljoscha are other places starting to adopt the style of having the name in the method directly? Then makes sense to do this here as well, I agree.

**Comment by SteNicholas** [comment]
@StephanEwen @aljoscha , IMO, the `fromSequence()` interface is just used to simply generate a data stream containing a sequence, no watermark strategy. Most data stream requirements that contain sequences only need this. If users need the sequence stream with user-defined watermark strategy, using `fromSource()` makes sense. I think these convenient methods like this is user-friendly for most users. And about the name, what is the final conclusion of this naming method?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 2 above)
Comment (SteNicholas) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:848]: @StephanEwen Does this need to mention that implementation detail? I thought this need be mentioned here.

**Comment by aljoscha** [comment]
@StephanEwen Only `fromSource()` and the older `addSource()` allow setting the name in the method. I think we can add more of that but I would prefer if the name comes first, as said above because I think it will make for more legible client code.

Maybe this got lost from above: "What we could do is use the Long itself as the timestamp and use it for watermarks. I think that would be the only sensible basis for a timestamp/watermark that is deterministic. What do you think?"

But I also like the idea of not having a convenience method at all.

**Comment by StephanEwen** [comment]
Okay, lets add the name into the method. So basically have `fromSequence("name", from to)`. That sounds good.

@aljoscha and @SteNicholas the one worry I have about not adding a variant with watermark strategy is that it cannot be used at all for streaming use cases, because in the future we don't have the case any more where watermarks are generated later in the stream.

Maybe let's think how a `fromElements()` method would look like. Would it have a WatermarkStrategy or not. I think that `fromSequence()` and `fromElemets()` should be symmetric there.

**Comment by SteNicholas** [comment]
> Okay, lets add the name into the method. So basically have `fromSequence("name", from to)`. That sounds good.
> 
> @aljoscha and @SteNicholas the one worry I have about not adding a variant with watermark strategy is that it cannot be used at all for streaming use cases, because in the future we don't have the case any more where watermarks are generated later in the stream.
> 
> Maybe let's think how a `fromElements()` method would look like. Would it have a WatermarkStrategy or not. I think that `fromSequence()` and `fromElemets()` should be symmetric there.

@StephanEwen For the name into the `fromSequence(sourceName, from to)`, I have followed your idea to update the name of the method.
About WatermarkStrategy, IMO, I still stick to the point that `fromSequence` and `fromElements()` method wouldn't have WatermarkStrategy, because what benefit could users get from if these methods have WatermarkStrategy? 
@aljoscha , any idea for how a `fromElements()` method would look like and would it have a WatermarkStrategy or not?

**Comment by aljoscha** [comment]
I think we would still have both variants for `fromSequence()`: one that takes a name and one that doesn't. Most users probably don't want to set a name immediately.

Also, we should then change `fromSource()` to also put the name argument first to be consistent.

I would table the `WatermarkStrategy` question for now. We can always add a method that takes on later, if we feel the need.

**Comment by aljoscha** [comment]
We should keep this as simple as possible for now and add more in case we need it later. I condensed it down to this commit and would merge that if no-one objects: https://github.com/apache/flink/commit/411b4c1e4ad88a7468343efa04dd87c12d0dc5f8.

**Comment by SteNicholas** [comment]
> We should keep this as simple as possible for now and add more in case we need it later. I condensed it down to this commit and would merge that if no-one objects: [411b4c1](https://github.com/apache/flink/commit/411b4c1e4ad88a7468343efa04dd87c12d0dc5f8).

@aljoscha , I agree with the commit and need @StephanEwen take a look.

**Comment by aljoscha** [comment]
@SteNicholas we shouldn't change the argument order because that would break existing code. That's why now the new `fromSequence()` doesn't have anything besides `from` and `to`. To keep it to a minimum.

Thanks for the comment on 411b4c1, I'm fixing it. 👌

I talked to Stephan myself, he's also fine with the changes.

**Comment by SteNicholas** [comment]
> @SteNicholas we shouldn't change the argument order because that would break existing code. That's why now the new `fromSequence()` doesn't have anything besides `from` and `to`. To keep it to a minimum.
> 
> Thanks for the comment on [411b4c1](https://github.com/apache/flink/commit/411b4c1e4ad88a7468343efa04dd87c12d0dc5f8), I'm fixing it. 👌
> 
> I talked to Stephan myself, he's also fine with the changes.

@aljoscha OK, I have no problem. Go ahead to merge this pull request.

**Review by aljoscha — APPROVED** [review]

**Comment by aljoscha** [comment]
I merged this. 👌

**Pull request closed without merging**